packages/@tailwindcss-upgrade/src/codemods/migrate-border-compatibility.ts

```diff
@@ -0,0 +1,200 @@
+import dedent from 'dedent'
+import postcss, { AtRule, type Plugin, type Root } from 'postcss'
+import type { Config } from 'tailwindcss'
+import { keyPathToCssProperty } from '../../../tailwindcss/src/compat/apply-config-to-theme'
+import type { DesignSystem } from '../../../tailwindcss/src/design-system'
+import { toKeyPath } from '../../../tailwindcss/src/utils/to-key-path'
+import * as ValueParser from '../../../tailwindcss/src/value-parser'
+
+// Defaults in v4
+const DEFAULT_BORDER_COLOR = 'currentColor'
+
+const css = dedent
+const COMPATIBILITY_CSS = css`
+  /**
+   * The default border styles in Tailwind CSS v4 changed compared to 
+   * Tailwind CSS v3. To make sure your project looks the same, we've added
+   * these compatibility styles to your project.
+   *
+   * It is recommended to migrate your project, and delete this layer once you
+   * are done.
+   *
+   * - All elements with a 'border' class will default to 'currentColor' in
+   *   Tailwind CSS v4, instead of 'border-gray-200' in Tailwind CSS v3.
+   * - All inputs, selects, and textareas will have a 'border-width: 1px' in
+   *   Tailwind CSS v4, instead of 'border-width: 0' in Tailwind CSS v3.
+   */
+  @layer base {
```

Comment:
In my testing the `@layer base` isn't strictly required (once the previous PR where we updated `preflight.css` is in place). But I think this is still nicer because:

1. It is about the reset that already lives in the `base` layer
2. It groups the compatibility CSS together, which is nice if you look at the CSS files we've updated